waspc/data/Generator/templates/server/src/auth/providers/oauth/state.ts

```diff
@@ -1,70 +1,93 @@
 import {
   Response as ExpressResponse,
   Request as ExpressRequest,
-} from "express";
-import { generateCodeVerifier, generateState } from "arctic";
+} from 'express';
+import { generateCodeVerifier, generateState } from 'arctic';
 
-import type { ProviderConfig } from "wasp/auth/providers/types";
+import type { ProviderConfig } from 'wasp/auth/providers/types';
 
-import { setOAuthCookieValue, getOAuthCookieValue } from "./cookies.js";
+import { setOAuthCookieValue, getOAuthCookieValue } from './cookies.js';
 
-export type StateType = 'state' | 'codeVerifier';
+/**
+ * OAuth state shape depends on whether the provider uses PKCE.
+ * If a provider uses PKCE, the state will include a code verifier.
+ */
+export type OAuthState<UsesCodeVerifier extends boolean = false> = {
+  state: string;
+} & (UsesCodeVerifier extends true
```

Comment:
Boolean type arguments are generally considered an antipattern (even more than boolean arguments in functions, who also get a bad rep). This is because they move the type system away from "declaring sets of values" and into "implementing decision algorithms":
 - `Foo<User>` - It's a `Foo` of Users (whatever `Foo` is).
 - `Foo<true>` - Hmm.
 
Of course, boolean type arguments are sometimes necessary, and we do use them throughout the codebase. But if we can help it, we should avoid them.

All in all, this is a pretty complex type, it includes: a type parameter (and a boolean one at that), a default value, a type constraint, an intersection type, and an inline conditional type.

I understand this code, but put yourself in someone else's shoes. Would you say the prevented duplication/extra keystrokes are worth all the additional complexity?

We do sometimes use all of these things in our generics, but we're

Reply:
Ok, I've decided to be explicit about the OAuth flow type we are using i.e. with or without PKCE and this is what will result in the state object type.

Reply:
Something like
```ts
export type OAuthStateFor<
  OT extends OAuthType
> = OAuthStateForOAuthType[OT];

export type OAuthStateWithCodeFor<OT extends OAuthType> = OAuthStateFor<OT> & {
  code: string;
};

export type OAuthType = keyof OAuthStateForOAuthType;

export type OAuthStateFieldName = keyof OAuthState | keyof OAuthStateWithPKCE;

type OAuthStateForOAuthType = {
  OAuth2: OAuthState,
  OAuth2WithPKCE: OAuthStateWithPKCE,
};

type OAuthState = {
  state: string;
};

type OAuthStateWithPKCE = {
  state: string;
  codeVerifier: string;
};
```